src/GraphQL/Internal/Validation.hs

```diff
@@ -64,15 +74,27 @@ data QueryDocument
   deriving (Eq, Show)
 
 data Operation
-  = Query [AST.VariableDefinition] Directives [Selection FragmentSpread]
-  | Mutation  [AST.VariableDefinition] Directives [Selection FragmentSpread]
+  = Query [AST.VariableDefinition] Directives SelectionSet
+  | Mutation  [AST.VariableDefinition] Directives SelectionSet
   deriving (Eq, Show)
 
+-- | Get the selection set for an operation.
+--
+-- TODO: This doesn't return the *actual* selection set we need, because it
+-- hasn't substituted variables or applied directives.
+getSelectionSet :: Operation -> SelectionSet
+getSelectionSet (Query _ _ ss) = ss
+getSelectionSet (Mutation _ _ ss) = ss
+
 -- | Type alias for 'Query' and 'Mutation' constructors of 'Operation'.
-type OperationType = [AST.VariableDefinition] -> Directives -> [Selection FragmentSpread] -> Operation
+type OperationType = [AST.VariableDefinition] -> Directives -> SelectionSet -> Operation
 
 newtype Operations = Operations (Map Name Operation) deriving (Eq, Show)
 
+type SelectionSet = [Selection]
```

Comment:
`SelectionSet` is actually ordered so the name isn't super great (I know the name is from the spec so let's keep it, just something I noticed)